web3c/contract_factory.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * @returns a constructor to create a web3.eth.Contract object. The `Contract` object
+ *          here has Oasis specific functionality, such as the ability to specify the
+ *          Oasis contract header upon deploy and the ability to retrieve contract
+ *          expiry from an instance of such a Contract. This is used in both confidential
+ *          and non confidential settings.
+ * @param   {Object} web3 the web3 object.
+ * @param   {Function?} providerFn is a block to execute returning the contract's provider.
+ *          The providerFn should take one argument, the options given upon invoking the
+ *          Contract constructor, for example { saveSession: false } for confidential
+ *          contracts.
+ */
+function make(web3, providerFn) {
+  let EthContract = web3.eth.Contract;
+  /**
+   * @param {Object} abi
+   * @param {String} address
+   * @param {Object} options
+   * @param {String} options.key The longterm key of the contract.
+   * @param {bool}   options.saveSession false to disable storing keys.
+   */
+  return function OasisEthContract(abi, address, options) {
+    let c = new EthContract(abi, address, options);
+
+    Object.assign(this, c);
+    this.__proto__ = c.__proto__;
+
+    // Object.DefineProperty's are not copied otherwise.
+    this.defaultAccount = c.constructor.defaultAccount;
+    this.defaultBlock = c.constructor.defaultBlock || 'latest';
+
+    c.setProvider.call(this, providerFn(options));
+
+    this.clone = () => {
+      return new OasisEthContract(this.options.jsonInterface, this.options.address, this.options);
+    };
+    // Hook deploy so that we can pass in the Oasis contract deployment header
+    // as an extra argument. For example, contract.deploy({ data, header: { expiry } });
+    // To do this, we need to also hook into all the methods available on the returned
+    // tx object so that we can pass in such deploy options into them.
+    this.deploy = (deployOptions, callback) => {
+      deployOptions = Object.assign({}, deployOptions);
+      // Create the txObject that we want to patch and return.
+      let txObject = c.deploy.call(this, deployOptions, callback);
+
+      // Methods we want to hook into.
+      let _send = txObject.send;
+      let _estimateGas = txObject.estimateGas;
+
+      // Perform patches.
+      txObject.send = (options) => {
+        options = Object.assign({}, options);
```

Comment:
why is this needed?

Reply:
In case the deploy options is undefined

Reply:
Let me see if I can remove this line. I recall some weirdness with the monkey patching.

Reply:
So this is needed because `options` is an optional argument. Sometimes it is undefined, in which case we need to replace it with {} so that we can give it the `deployOptions.header` in the next line below.

Reply:
`options = options || {};`  could be a simpler way to accomplish that?